.bash_prompt

```diff
@@ -14,54 +14,16 @@ elif infocmp xterm-256color >/dev/null 2>&1; then
 fi;
 
 prompt_git() {
-	local s='';
-	local branchName='';
-
-	# Check if the current directory is in a Git repository.
-	if [ "$(git rev-parse --is-inside-work-tree &>/dev/null; echo "${?}")" == '0' ]; then
-
-		# check if the current directory is in .git before running git checks
-		if [ "$(git rev-parse --is-inside-git-dir 2> /dev/null)" == 'false' ]; then
-
-			if [[ -O "$(git rev-parse --show-toplevel)/.git/index" ]]; then
-				git update-index --really-refresh -q &> /dev/null;
-			fi;
-
-			# Check for uncommitted changes in the index.
-			if ! git diff --quiet --ignore-submodules --cached; then
-				s+='+';
-			fi;
-
-			# Check for unstaged changes.
-			if ! git diff-files --quiet --ignore-submodules --; then
-				s+='!';
-			fi;
-
-			# Check for untracked files.
-			if [ -n "$(git ls-files --others --exclude-standard)" ]; then
-				s+='?';
-			fi;
-
-			# Check for stashed files.
-			if git rev-parse --verify refs/stash &>/dev/null; then
-				s+='$';
-			fi;
-
-		fi;
-
-		# Get the short symbolic ref.
-		# If HEAD isn’t a symbolic ref, get the short SHA for the latest commit
-		# Otherwise, just give up.
-		branchName="$(git symbolic-ref --quiet --short HEAD 2> /dev/null || \
-			git rev-parse --short HEAD 2> /dev/null || \
-			echo '(unknown)')";
-
-		[ -n "${s}" ] && s=" [${s}]";
-
-		echo -e "${1}${branchName}${blue}${s}";
-	else
-		return;
-	fi;
+    local gs='' s='' branch='';
+    gs=$(git status --show-stash 2>&1) || return;
+    branch=$(echo "$gs" | head -n1)
+    branch=${branch#"On branch "}
```

Comment:
This would be faster, as it's pure bash:
```
  re='On branch ([^'$'\n'']+)'
  if [[ "$gs" =~ $re ]]; then
    branch=${BASH_REMATCH[1]}
  else
    branch=''
  fi

Reply:
Thanks @rasa . It may be faster, but I think the other is a tad clearer, and we're also probably talking a very small timing difference. Do you have numbers showing otherwise?

Reply:
Another pure-bash (v4) option would be:
```sh
mapfile -tn 1 <<<"$gs"
branch=${MAPFILE[0]#"On branch "}
```
Both alternatives don’t execute a sub-shell, or an external program (head), so they must be faster. I can write a script to test the differences.

Reply:
Based on the results below, the pure-bash regex solution (test 4) is > 99% faster:
```
test1 (echo | head):
real    0m9.911s
user    0m1.000s
sys     0m9.344s

test1 (head <<<):
real    0m9.903s
user    0m0.688s
sys     0m8.859s

test3 (mapfile):
real    0m0.855s
user    0m0.047s
sys     0m0.813s

test4 (regex):
real    0m0.023s
user    0m0.031s
sys     0m0.000s
```
Using this test script:
```bash
#!/usr/bin/env bash

read -r -d '' gs <<'EOF'
On branch master
Your branch is up to date with 'origin/master'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

no changes added to commit (use "git add" and/or "git commit -a")
EOF

tries=1000

test1() {
  for (( i = 1; i <= $tries ; i++ )) {
    branch=$(echo "$gs" | head -n1)
    branch=${branch#"On branch "}
  }
}

test2() {
  for (( i = 1; i <= $tries ; i++ )) {
    branch=$(head -n1 <<<"$gs")
    branch=${branch#"On branch "}
  }
}

test3() {
  for (( i = 1; i <= $tries ; i++ )) {
    mapfile -tn 1 <<<"$gs"
    branch=${MAPFILE[0]#'On branch '}
  }
}

test4() {
  for (( i = 1; i <= $tries ; i++ )) {
    re='On branch ([^'$'\n'']+)'
    if [[ "$gs" =~ $re ]]; then
      branch=${BASH_REMATCH[1]}
    else
      branch=''
    fi
  }
}

printf '\ntest1 (echo | head):'
time test1
printf '\ntest1 (head <<<):'
time test2
printf '\ntest3 (mapfile):'
time test3
printf '\ntest4 (regex):'
time test4
```

Reply:
This also works, and is about 10% faster:
```
    [[ "$gs" =~ On\ branch\ ([[:print:]]*) ]] && branch=${BASH_REMATCH[1]}
```

Reply:
Thanks @rasa for those numbers. Given that's over 1000 tries, we're talking about 10ms, which is pretty significant given it's taking 30ms all up (i.e. this will drop it down to 20ms). Thanks! Will add in your last one.

Reply:
@kodonell And that’s on fast hardware. On Pis, OpenWrt’s, etc, these tweaks are even more significant.